### PR TITLE
8264054: Bad XMM performance on java.lang.MathBench.sqrtDouble

### PR DESCRIPTION
**Pull request by sviswa7**
For the j.l.Math JMH at https://github.com/openjdk/jmh-jdk-microbenchmarks/blob/master/micros-jdk11/src/main/java/org/openjdk/bench/java/lang/MathBench.java, the performance for sqrt benchmark could be improved.  Thanks a lot to Eric Caspole for finding this issue.

Benchmark:
    @Benchmark
    public double  sqrtDouble() {
        return  Math.sqrt(double4Dot1);
    }

Current code generated (linux format) by c2 JIT is:
     vsqrtsd 0x50(%r10),%xmm0,%xmm0

The vsqrtsd instruction operation is specified as below: 
     VSQRTSD (VEX.128 encoded version)
     DEST[63:0] := SQRT(SRC2[63:0])
     DEST[127:64] := SRC1[127:64]
     DEST[MAXVL-1:128] := 0

The upper 127:64 bits are set from previous contents of xmm0. As the destination xmm0 register was not initialized prior to use by c2 JIT, this causes stall and lower performance.

By adding xmm0 initialization prior to use, the performance of the above benchmark improves significantly.

Code generated after patch:
     vxorpd %xmm0,%xmm0,%xmm0
     vsqrtsd 0x50(%r10),%xmm0,%xmm0

Performance before patch:
Benchmark                    Mode  Cnt       Score      Error   Units
MathBench.sqrtDouble        thrpt    8  193612.396 ±  95.807  ops/ms

Performance after patch:
MathBench.sqrtDouble        thrpt    8  276388.024 ± 846.372  ops/ms

Best Regards,
Sandhya

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264054](https://bugs.openjdk.java.net/browse/JDK-8264054): Bad XMM performance on java.lang.MathBench.sqrtDouble

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 2cfa863b4fae157be5ae40e5a67aec7ecef63eb9
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 2cfa863b4fae157be5ae40e5a67aec7ecef63eb9
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 2cfa863b4fae157be5ae40e5a67aec7ecef63eb9

### Contributors
 * Eric Caspole `<ecaspole@openjdk.org>`
 * Charlie Hunt `<huntch@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3256/head:pull/3256` \
`$ git checkout pull/3256`

Update a local copy of the PR: \
`$ git checkout pull/3256` \
`$ git pull https://git.openjdk.java.net/jdk pull/3256/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3256`

View PR using the GUI difftool: \
`$ git pr show -t 3256`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3256.diff">https://git.openjdk.java.net/jdk/pull/3256.diff</a>

</details>

### CODE REVIEW

**Review by neliasso — COMMENTED** [review]
In the bug report you are mentioning new test micro benchmark test cases. Have you already contributed those, or would you like to add those to this PR?

**Comment by ericcaspole** [comment]
I added these micros the the  https://github.com/openjdk/jmh-jdk-microbenchmarks a couple weeks ago, before we knew about this problem, because it is easy to work on JMH with Maven.  But it would be best to contribute these micros into the JDK repo at the same time as this patch. I will get it together for the JDK repo and connect with Sandhya.

**Review by vnkozlov — COMMENTED** [review]
What is faster: `xor(xmm) + sqrt(xmm, mem)` or `mov(xmm, mem) + sqrt(xmm,xmm)` ?

Do we have other instructions which may suffer stall too if `dst` register is not zeroed?

File: src/hotspot/cpu/x86/x86.ad [hunk 1]
```diff
@@ -3247,10 +3249,11 @@ instruct sqrtF_reg(regF dst, regF src) %{
 instruct sqrtF_mem(regF dst, memory src) %{
   predicate(UseSSE>=1);
   match(Set dst (SqrtF (LoadF src)));
-
+  effect(TEMP dst);
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:3252]: Why do you declare `dst` as `TEMP` (here and in other places)?

**Comment by sviswa7** [comment]
@vnkozlov  Both xor(xmm) + sqrt(xmm, mem) or mov(xmm, mem) + sqrt(xmm,xmm) have same performance. I could look into simplifying the patch by just keeping the register version of these rules.

I looked through x86.ad for other usages of unary instructions but didn't find any other instances. We implement negate and abs using binary instructions.

File: src/hotspot/cpu/x86/x86.ad (hunk 1 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:3252]: Good point. I was being extra cautious that the register in address should not be overwritten by the xorps. But of course the address cannot have xmm register here so TEMP dst is not needed.  I will update the patch accordingly.

**Comment by sviswa7** [comment]
@neliasso @vnkozlov @iwanowww All the review comments are implemented.  
The MathBench.java and StrictMathBench.java contribution is from Eric Caspole and Charlie Hunt.
Please let me know if any other changes are needed.

**Review by vnkozlov — APPROVED** [review]
Nice solution - let C2 generate the best value loading instruction. I assume the improvement is the same with this update.

Thank you for checking other instructions.

**Comment by sviswa7** [comment]
@vnkozlov Yes the performance improvement is the same with the updated patch.

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -3232,73 +3232,22 @@ instruct negD_reg_reg(vlRegD dst, vlRegD src) %{
   ins_pipe(pipe_slow);
 %}
 
-instruct sqrtF_reg(regF dst, regF src) %{
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86.ad:3235]: Would be helpful to have a comment describing why there are only reg-to-reg variants kept for `SqrtF`/`SqrtD`.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by neliasso — APPROVED** [review]
Thanks for adding the test.

Looks good.

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:3235]: Done, added comments for the sqrt rules.

**Pull request closed without merging**